File: android/src/com/marpies/ane/onesignal/functions/PostNotificationFunction.java
/*
 * Copyright 2016 Marcel Piestansky (http://marpies.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.marpies.ane.onesignal.functions;

import com.adobe.fre.FREContext;
import com.adobe.fre.FREObject;
import com.marpies.ane.onesignal.data.OneSignalEvent;
import com.marpies.ane.onesignal.utils.AIR;
import com.marpies.ane.onesignal.utils.FREObjectUtils;
import com.onesignal.OneSignal;
import org.json.JSONException;
import org.json.JSONObject;

public class PostNotificationFunction extends BaseFunction {

	@Override
	public FREObject call( FREContext context, FREObject[] args ) {
		super.call( context, args );

		String parametersString = FREObjectUtils.getString( args[0] );
		final int callbackID = FREObjectUtils.getInt( args[1] );
		AIR.log( "OneSignal::postNotification" );

		try {
			JSONObject parameters = new JSONObject( parametersString );
			OneSignal.postNotification( parameters, new OneSignal.PostNotificationResponseHandler() {
				@Override
				public void onSuccess( JSONObject jsonObject ) {
					AIR.log( "OneSignal::postNotification | success" );
					JSONObject response = new JSONObject();
					try {
						response.put( "callbackID", callbackID );
						response.put( "successResponse", jsonObject.toString() );
						AIR.dispatchEvent( OneSignalEvent.POST_NOTIFICATION_SUCCESS, response.toString() );
					} catch( JSONException e ) {
						e.printStackTrace();
					}
				}

				@Override
				public void onFailure( JSONObject jsonObject ) {
					AIR.log( "OneSignal::postNotification | error" );
					dispatchFailure( callbackID, jsonObject );
				}
			} );
		} catch( Exception e ) {
			e.printStackTrace();
			try {
				JSONObject jsonObject = new JSONObject();
				jsonObject.put( "error", e.getLocalizedMessage() );
				dispatchFailure( callbackID, jsonObject );
			} catch( Exception e1 ) {
				e1.printStackTrace();
			}
		}

		return null;
	}

	private void dispatchFailure( final int callbackID, JSONObject jsonObject ) {
		JSONObject response = new JSONObject();
		try {
			response.put( "callbackID", callbackID );
			response.put( "errorResponse", jsonObject.toString() );
			AIR.dispatchEvent( OneSignalEvent.POST_NOTIFICATION_ERROR, response.toString() );
		} catch( JSONException e ) {
			e.printStackTrace();
		}
	}

}

File: ios/OneSignal/Functions/PostNotificationFunction.m
/**
 * Copyright 2016 Marcel Piestansky (http://marpies.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "PostNotificationFunction.h"
#import <AIRExtHelpers/MPFREObjectUtils.h>
#import <AIRExtHelpers/MPStringUtils.h>
#import "AIROneSignal.h"
#import "OneSignalEvent.h"

FREObject pushos_postNotification( FREContext context, void* functionData, uint32_t argc, FREObject argv[] ) {
    [AIROneSignal log:@"pushos_postNotification"];
    NSString* parametersString = [MPFREObjectUtils getNSString:argv[0]];
    int callbackID = [MPFREObjectUtils getInt:argv[1]];
    
    NSError* jsonError = [[NSError alloc] init];
    NSData* jsonData = [parametersString dataUsingEncoding:NSUTF8StringEncoding];
    NSDictionary* jsonResult = [NSJSONSerialization JSONObjectWithData:jsonData
                                                         options:NSJSONReadingMutableContainers
                                                           error:&jsonError];
    /* String parsed to JSON successfully */
    if( jsonResult != nil ) {
        [[[AIROneSignal sharedInstance] appDelegate] postNotification:jsonResult callbackID:callbackID];
    }
    /* There was an error parsing the JSON String */
    else {
        NSMutableDictionary* errorResponse = [NSMutableDictionary dictionary];
        errorResponse[@"callbackID"] = [NSNumber numberWithInt:callbackID];
        errorResponse[@"errorResponse"] = @{ @"error": jsonError.localizedDescription };
        [AIROneSignal dispatchEvent:POST_NOTIFICATION_ERROR withMessage:[MPStringUtils getJSONString:errorResponse]];
    }
    return nil;
}